File: apps/backend/internal/entity/exam.go
package entity

import (
	"time"

	"github.com/google/uuid"
)

// ExamStatus represents the status of an exam
// Aligned with ExamSystem.md design and Question system
type ExamStatus string

const (
	ExamStatusActive   ExamStatus = "ACTIVE"   // Đã xuất bản, students có thể làm (was 'published')
	ExamStatusPending  ExamStatus = "PENDING"  // Đang soạn thảo, chờ review (was 'draft')
	ExamStatusInactive ExamStatus = "INACTIVE" // Tạm ngưng (new status)
	ExamStatusArchived ExamStatus = "ARCHIVED" // Đã lưu trữ (was 'archived')
)

// ExamType represents the type of exam
// Aligned with ExamSystem.md design - simplified to focus on source distinction
type ExamType string

const (
	ExamTypeGenerated ExamType = "generated" // Đề thi tạo từ ngân hàng câu hỏi
	ExamTypeOfficial  ExamType = "official"  // Đề thi thật từ trường/sở
)

// Difficulty represents the difficulty level of an exam
// Aligned with Question system difficulty enum
type Difficulty string

const (
	DifficultyEasy   Difficulty = "EASY"   // Dễ
	DifficultyMedium Difficulty = "MEDIUM" // Trung bình
	DifficultyHard   Difficulty = "HARD"   // Khó
	DifficultyExpert Difficulty = "EXPERT" // Rất khó (aligned with Question system)
)

// Exam represents an exam definition
// Updated to align with ExamSystem.md design
type Exam struct {
	ID              string     `json:"id" db:"id"`
	Title           string     `json:"title" db:"title"`
	Description     string     `json:"description" db:"description"`
	Instructions    string     `json:"instructions" db:"instructions"`
	DurationMinutes int        `json:"duration_minutes" db:"duration_minutes"`
	TotalPoints     int        `json:"total_points" db:"total_points"`
	PassPercentage  int        `json:"pass_percentage" db:"pass_percentage"`
	ExamType        ExamType   `json:"exam_type" db:"exam_type"`
	Status          ExamStatus `json:"status" db:"status"`

	// Academic Classification (ESSENTIAL - from ExamSystem.md)
	Subject    string     `json:"subject" db:"subject"`       // Môn học (REQUIRED)
	Grade      *int       `json:"grade" db:"grade"`           // Khối lớp (1-12)
	Difficulty Difficulty `json:"difficulty" db:"difficulty"` // EASY, MEDIUM, HARD, EXPERT
	Tags       []string   `json:"tags" db:"tags"`             // Tags tìm kiếm

	// Settings
	ShuffleQuestions bool `json:"shuffle_questions" db:"shuffle_questions"`
	ShowResults      bool `json:"show_results" db:"show_results"`
	MaxAttempts      int  `json:"max_attempts" db:"max_attempts"`

	// Official Exam Fields (OPTIONAL - chỉ cho exam_type = 'official')
	SourceInstitution *string `json:"source_institution,omitempty" db:"source_institution"` // Tên trường/sở
	ExamYear          *string `json:"exam_year,omitempty" db:"exam_year"`                   // Năm thi (VD: "2024")
	ExamCode          *string `json:"exam_code,omitempty" db:"exam_code"`                   // Mã đề (VD: "001", "A")
	FileURL           *string `json:"file_url,omitempty" db:"file_url"`                     // Link file PDF

	// Integration Fields (NEW)
	Version int `json:"version" db:"version"` // For optimistic locking

	// Questions (loaded separately)
	QuestionIDs []string `json:"question_ids"`

	// Timestamps
	CreatedBy   string     `json:"created_by" db:"created_by"`
	PublishedAt *time.Time `json:"published_at" db:"published_at"`
	CreatedAt   time.Time  `json:"created_at" db:"created_at"`
	UpdatedAt   time.Time  `json:"updated_at" db:"updated_at"`
}

// AttemptStatus represents the status of an exam attempt
type AttemptStatus string

const (
	AttemptStatusInProgress AttemptStatus = "in_progress"
	AttemptStatusSubmitted  AttemptStatus = "submitted"
	AttemptStatusGraded     AttemptStatus = "graded"
	AttemptStatusCancelled  AttemptStatus = "cancelled"
)

// ExamAttempt represents a user's attempt at an exam
type ExamAttempt struct {
	ID            string        `json:"id" db:"id"`
	ExamID        string        `json:"exam_id" db:"exam_id"`
	UserID        string        `json:"user_id" db:"user_id"`
	AttemptNumber int           `json:"attempt_number" db:"attempt_number"`
	Status        AttemptStatus `json:"status" db:"status"`
	Score         int           `json:"score" db:"score"`
	TotalPoints   int           `json:"total_points" db:"total_points"`
	Percentage    float64       `json:"percentage" db:"percentage"`
	Passed        bool          `json:"passed" db:"passed"`
	StartedAt     time.Time     `json:"started_at" db:"started_at"`
	SubmittedAt   *time.Time    `json:"submitted_at" db:"submitted_at"`
	TimeSpent     int           `json:"time_spent_seconds" db:"time_spent_seconds"`
}

// ExamQuestion represents a question in an exam
type ExamQuestion struct {
	ID          string    `json:"id" db:"id"`
	ExamID      string    `json:"exam_id" db:"exam_id"`
	QuestionID  string    `json:"question_id" db:"question_id"`
	OrderNumber int       `json:"order_number" db:"order_number"`
	Points      int       `json:"points" db:"points"`
	IsBonus     bool      `json:"is_bonus" db:"is_bonus"`
	CreatedAt   time.Time `json:"created_at" db:"created_at"`
}

// ExamAnswer represents a user's answer to a question in an exam attempt
type ExamAnswer struct {
	ID               string    `json:"id" db:"id"`
	AttemptID        string    `json:"attempt_id" db:"attempt_id"`
	QuestionID       string    `json:"question_id" db:"question_id"`
	AnswerData       string    `json:"answer_data" db:"answer_data"` // JSONB stored as string
	IsCorrect        *bool     `json:"is_correct" db:"is_correct"`   // Nullable for manual grading
	PointsEarned     float64   `json:"points_earned" db:"points_earned"`
	TimeSpentSeconds *int      `json:"time_spent_seconds" db:"time_spent_seconds"`
	AnsweredAt       time.Time `json:"answered_at" db:"answered_at"`
}

// ExamResult represents summary statistics for a completed exam attempt
type ExamResult struct {
	ID                 string    `json:"id" db:"id"`
	AttemptID          string    `json:"attempt_id" db:"attempt_id"`
	TotalQuestions     int       `json:"total_questions" db:"total_questions"`
	CorrectAnswers     int       `json:"correct_answers" db:"correct_answers"`
	IncorrectAnswers   int       `json:"incorrect_answers" db:"incorrect_answers"`
	Unanswered         int       `json:"unanswered" db:"unanswered"`
	ScoreBreakdown     string    `json:"score_breakdown" db:"score_breakdown"` // JSONB stored as string
	AccuracyPercentage *float64  `json:"accuracy_percentage" db:"accuracy_percentage"`
	AvgTimePerQuestion *float64  `json:"avg_time_per_question" db:"avg_time_per_question"`
	Feedback           *string   `json:"feedback" db:"feedback"`
	Grade              *string   `json:"grade" db:"grade"` // A+, A, B+, B, C, D, F
	CreatedAt          time.Time `json:"created_at" db:"created_at"`
}

// ExamFeedback represents user feedback about an exam
type ExamFeedback struct {
	ID               string    `json:"id" db:"id"`
	ExamID           string    `json:"exam_id" db:"exam_id"`
	UserID           string    `json:"user_id" db:"user_id"`
	AttemptID        *string   `json:"attempt_id" db:"attempt_id"`
	Rating           *int      `json:"rating" db:"rating"`                       // 1-5 scale
	DifficultyRating *int      `json:"difficulty_rating" db:"difficulty_rating"` // 1-5 scale
	Content          *string   `json:"content" db:"content"`
	CreatedAt        time.Time `json:"created_at" db:"created_at"`
}

// NewExam creates a new exam with defaults
// Updated to align with ExamSystem.md design
func NewExam(title, description string, createdBy string) *Exam {
	return &Exam{
		ID:              uuid.New().String(),
		Title:           title,
		Description:     description,
		DurationMinutes: 60,
		PassPercentage:  60,
		ExamType:        ExamTypeGenerated, // Default to 'generated'
		Status:          ExamStatusPending, // Default to 'PENDING'

		// Academic Classification defaults
		Subject:    "",               // Must be set by caller
		Grade:      nil,              // Optional
		Difficulty: DifficultyMedium, // Default difficulty
		Tags:       []string{},

		// Settings defaults
		ShuffleQuestions: false,
		ShowResults:      true,
		MaxAttempts:      1,

		// Official exam fields (nil for generated exams)
		SourceInstitution: nil,
		ExamYear:          nil,
		ExamCode:          nil,
		FileURL:           nil,

		// Integration fields
		Version: 1, // Start with version 1

		// Questions
		QuestionIDs: []string{},

		// Timestamps
		CreatedBy: createdBy,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}
}

File: apps/frontend/src/types/exam.ts
/**
 * Exam Types for NyNus Exam System
 * Interfaces và types cho quản lý exam system
 * Aligned với ExamSystem.md design và backend entities
 * 
 * @author NyNus Team
 * @version 1.0.0
 */

import { QuestionDifficulty, Question } from './question';

// Re-export Question for convenience
export type { Question };

// ===== ENUMS =====

/**
 * Exam Status Enum
 * Aligned với backend ExamStatus và protocol buffers
 */
export enum ExamStatus {
  ACTIVE = 'ACTIVE',       // Đã xuất bản, students có thể làm
  PENDING = 'PENDING',     // Đang soạn thảo, chờ review
  INACTIVE = 'INACTIVE',   // Tạm ngưng
  ARCHIVED = 'ARCHIVED'    // Đã lưu trữ
}

/**
 * Exam Type Enum
 * Aligned với backend ExamType
 */
export enum ExamType {
  GENERATED = 'GENERATED', // Đề thi tạo từ ngân hàng câu hỏi
  OFFICIAL = 'OFFICIAL'    // Đề thi thật từ trường/sở
}

/**
 * Attempt Status Enum
 * Aligned với backend AttemptStatus và protocol buffers
 */
export enum AttemptStatus {
  IN_PROGRESS = 'IN_PROGRESS', // Đang làm bài
  SUBMITTED = 'SUBMITTED',     // Đã nộp bài
  GRADED = 'GRADED',          // Đã chấm điểm
  CANCELLED = 'CANCELLED'      // Đã hủy
}

// ===== CORE INTERFACES =====

/**
 * Main Exam Interface
 * Aligned với backend entity/exam.go
 */
export interface Exam {
  // Basic Information
  id: string;
  title: string;
  description: string;
  instructions: string;
  durationMinutes: number;
  totalPoints: number;
  passPercentage: number;
  examType: ExamType;
  status: ExamStatus;

  // Academic Classification (ESSENTIAL)
  subject: string;           // Môn học (REQUIRED)
  grade?: number;           // Khối lớp (1-12)
  difficulty: QuestionDifficulty; // EASY, MEDIUM, HARD, EXPERT
  tags: string[];           // Tags tìm kiếm

  // Settings
  shuffleQuestions: boolean;
  showResults: boolean;
  maxAttempts: number;

  // Official Exam Fields (OPTIONAL - chỉ cho examType = 'OFFICIAL')
  sourceInstitution?: string; // Tên trường/sở
  examYear?: string;         // Năm thi (VD: "2024")
  examCode?: string;         // Mã đề (VD: "001", "A")
  fileUrl?: string;          // Link file PDF

  // Integration Fields
  version: number;           // For optimistic locking

  // Questions (loaded separately)
  questionIds: string[];

  // Timestamps
  createdAt: string;
  updatedAt: string;
  createdBy: string;
  updatedBy?: string;

  // Relations (optional, loaded when needed)
  creator?: {
    id: string;
    name: string;
    email: string;
  };
  examResults?: ExamResult[];
  examQuestions?: ExamQuestion[];
}

/**
 * Exam Attempt Interface
 * Represents a user's attempt at an exam
 */
export interface ExamAttempt {
  id: string;
  examId: string;
  userId: string;
  attemptNumber: number;
  status: AttemptStatus;
  score?: number;
  totalPoints: number;
  percentage?: number;
  passed?: boolean;
  startedAt: string;
  submittedAt?: string;
  completedAt?: string;
  timeSpentSeconds?: number;

  // Additional data
  ipAddress?: string;
  userAgent?: string;
  notes?: string;

  // Relations (optional)
  exam?: Exam;
  answers?: ExamAnswer[];
  result?: ExamResult;
}

/**
 * Exam Answer Interface
 * Represents a user's answer to a question in an exam attempt
 */
export interface ExamAnswer {
  id: string;
  attemptId: string;
  questionId: string;
  answerData: string;        // JSONB stored as string
  isCorrect?: boolean;       // Nullable for manual grading
  pointsEarned: number;
  timeSpentSeconds?: number;
  answeredAt: string;
}

/**
 * Exam Result Interface
 * Represents the final result of an exam attempt
 */
export interface ExamResult {
  id: string;
  attemptId: string;
  examId: string;
  userId: string;
  score: number;
  totalPoints: number;
  percentage: number;
  passed: boolean;
  gradedAt: string;
  gradedBy?: string;

  // Detailed breakdown
  correctAnswers: number;
  incorrectAnswers: number;
  unansweredQuestions: number;
  timeSpentSeconds: number;

  // Feedback
  feedback?: string;
  teacherNotes?: string;

  // Relations (optional)
  attempt?: ExamAttempt;
  exam?: Exam;
}

/**
 * Exam Question Interface
 * Represents the relationship between exam and questions
 */
export interface ExamQuestion {
  id: string;
  examId: string;
  questionId: string;
  order: number;
  points: number;
  required: boolean;

  // Question data (loaded when needed)
  question?: {
    id: string;
    content: string;
    type: string;
    difficulty: QuestionDifficulty;
  };
}

// ===== UTILITY TYPES =====

/**
 * Exam Form Data Interface
 * For create/edit exam forms
 */
export interface ExamFormData {
  title: string;
  description: string;
  instructions: string;
  durationMinutes: number;
  totalPoints: number;
  passPercentage: number;
  examType: ExamType;
  status: ExamStatus;

  // Academic Classification
  subject: string;
  grade?: number;
  difficulty: QuestionDifficulty;
  tags: string[];

  // Settings
  shuffleQuestions: boolean;
  shuffleAnswers?: boolean;
  showResults: boolean;
  showAnswers?: boolean;
  allowReview?: boolean;
  maxAttempts: number;

  // Official Exam Fields (conditional)
  sourceInstitution?: string;
  examYear?: string;
  examCode?: string;
  fileUrl?: string;

  // Questions
  questionIds: string[];
}

/**
 * Exam Answer Input Interface
 * Custom interface for exam taking
 */
export interface ExamAnswerInput {
  questionId: string;
  questionType: string;
  selectedOptions: string[];
  answerText: string;
  isFlagged: boolean;
}

/**
 * Exam Filters Interface
 * For search and filter functionality
 */
export interface ExamFilters {
  search?: string;
  status?: ExamStatus[];
  examType?: ExamType[];
  subject?: string[];
  grade?: number[];
  difficulty?: QuestionDifficulty[];
  tags?: string[];
  createdBy?: string[];
  dateRange?: {
    from: string;
    to: string;
  };
  // Pagination
  page?: number;
  limit?: number;
}

/**
 * Exam Sort Options
 * For sorting exam lists
 */
export interface ExamSortOptions {
  field: 'title' | 'createdAt' | 'updatedAt' | 'status' | 'difficulty' | 'totalPoints' | 'durationMinutes';
  direction: 'asc' | 'desc';
}

/**
 * Exam Statistics Interface
 * For exam analytics and reporting
 */
export interface ExamStatistics {
  examId: string;
  totalAttempts: number;
  completedAttempts: number;
  averageScore: number;
  passRate: number;
  averageTimeSpent: number;
  difficultyDistribution: {
    [key in QuestionDifficulty]: number;
  };
  scoreDistribution: {
    range: string;
    count: number;
  }[];
}

// ===== TYPE GUARDS =====

/**
 * Type guard to check if exam is official
 */
export function isOfficialExam(exam: Exam): boolean {
  return exam.examType === ExamType.OFFICIAL;
}

/**
 * Type guard to check if exam is published
 */
export function isPublishedExam(exam: Exam): boolean {
  return exam.status === ExamStatus.ACTIVE;
}

/**
 * Type guard to check if attempt is completed
 */
export function isCompletedAttempt(attempt: ExamAttempt): boolean {
  return attempt.status === AttemptStatus.SUBMITTED || 
         attempt.status === AttemptStatus.GRADED;
}
